src/ppx_deriving_qcheck/ppx_deriving_qcheck.ml

```diff
@@ -381,66 +351,123 @@ and gen_from_arrow ~loc ~env left right =
     | x -> (gen_from_type ~loc ~env x, [%expr o_nil])
   in
   let x, obs = aux right in
+  (* TODO: export this in qcheck_generators for https://github.com/c-cube/qcheck/issues/190 *)
   let arb = [%expr QCheck.make [%e x]] in
   [%expr
     QCheck.fun_nary QCheck.Tuple.([%e observable left] @-> [%e obs]) [%e arb]
     |> QCheck.gen]
 
-let gen_from_kind_variant ~loc ~env typ_name xs =
-  let (module A) = Ast_builder.make loc in
-  let is_rec (constr : constructor_declaration) : bool =
-    match constr.pcd_args with
-    | Pcstr_tuple xs -> List.exists (is_rec_typ typ_name) xs
-    | _ -> false
-  in
-  sized ~loc ~env typ_name is_rec (gen_from_constr ~loc) xs
+(** [gen_from_type_declaration loc td] creates a generator from the type declaration.
 
-let rec curry_args ~loc args body =
-  match args with
-  | [] -> body
-  | x :: xs -> [%expr fun [%p x] -> [%e curry_args ~loc xs body]]
+    It returns either `Recursive or `Normal.
 
-let gen_from_type_declaration ~loc ?(env = TypeGen.empty) td =
-  let name = td.ptype_name.txt in
-  let pat_gen = pat ~loc name in
+    - `Normal of expression:
+    The derived generator is not recursive, we return only the generator.
+
+    - `Recursive of expression * expression
+    The derived generator was recursive (i.e. val gen : n -> t Gen.t), we return
+    the sized generator version, and a normal generator using this last with
+    [Gen.sized].
+*)
+let gen_from_type_declaration ~loc ~env td =
+  let (module A) = Ast_builder.make loc in
+  let ty = env.curr_type in
+  let is_rec = List.mem env.curr_type env.rec_types in
```

Comment:
Since you already bind the curr_type field above, this can be written as:
```ocaml
let is_rec = List.mem ty env.rec_types in
```

Reply:
Yep, fd6821fd249e2fab9ce4b75f083e02278337a4dd